---
title: "ADEs and Agricolae Soils"
author: "Anderson Freitas - Luis F. Zagatto"
date: "Jul/14/2022"
output: html_notebook
---

How we made our analysis and figures:

## Dirty work
Here we will put all the libraries we needed for downstream analysis. I will use it to simplify and unify all libraries in one only chunk.

```{r message=FALSE}

#Libraries

library(hues)
library(knitr)
library(readr)
library(vegan)
library(dplyr)
library(ggpubr)
library(ggplot2)
library(magrittr)
library(microeco)
library(corrplot)
library(phyloseq)
library(agricolae)
library(tidyverse)
library(microbiome)
library(factoextra)

#############################################

summarySE <- function(data=NULL, measurevar, groupvars=NULL, na.rm=FALSE,
                      conf.interval=.95, .drop=TRUE) {
  library(plyr)
  
  # New version of length which can handle NA's: if na.rm==T, don't count them
  length2 <- function (x, na.rm=FALSE) {
    if (na.rm) sum(!is.na(x))
    else       length(x)
  }
  
  # This does the summary. For each group's data frame, return a vector with
  # N, mean, and sd
  datac <- ddply(data, groupvars, .drop=.drop,
                 .fun = function(xx, col) {
                   c(N    = length2(xx[[col]], na.rm=na.rm),
                     mean = mean   (xx[[col]], na.rm=na.rm),
                     sd   = sd     (xx[[col]], na.rm=na.rm)
                   )
                 },
                 measurevar
  )
  
  # Rename the "mean" column    
  datac <- plyr::rename(datac, c("mean" = measurevar))
  
  datac$se <- datac$sd / sqrt(datac$N)  # Calculate standard error of the mean
  
  # Confidence interval multiplier for standard error
  # Calculate t-statistic for confidence interval: 
  # e.g., if conf.interval is .95, use .975 (above/below), and use df=N-1
  ciMult <- qt(conf.interval/2 + .5, datac$N-1)
  datac$ci <- datac$se * ciMult
  
  return(datac)
}

#############################################

```

## Setting some questions of beauty

```{r}

theme_set(theme_bw())
pal = "Set3"
scale_colour_discrete <-  function(palname=pal, ...){
  scale_colour_brewer(palette=palname, ...)
}
scale_fill_discrete <-  function(palname=pal, ...){
  scale_fill_brewer(palette=palname, ...)
}

ze=c("#79b643", "#bb59bd", "#52ac7a", "#cf426b", "#4dadcf","#cf5234",
     "#7475cb", "#cea13d", "#be6c91", "#748037", "#be774e")

```

## Starting with non-genomic data

At a first moment, we will add only the mapping file to proceed to initial ande descriptive analysis of the data.

```{r}

map_tpsa <- read_delim("E:/Anderson-BackUp/TP_SA/map_tpsa.tsv", delim = "\t",
                       escape_double = FALSE, na = "NA", trim_ws = TRUE)
map = data.frame(map_tpsa)

```


## Importing files from dada2 and creating a phyloseq object

```{r}

setwd("E:/Anderson-BackUp/TP_SA/analysis")

seqtab = readRDS("dada2/seqtab_final.rds")
taxa = readRDS("dada2/taxonomy_genera.rds")
mapfile <- "E:/Anderson-BackUp/TP_SA/map_tpsa.csv"

map = import_qiime_sample_data(mapfile)

# Building a phyloseq object

ps <- phyloseq(tax_table(taxa), otu_table(seqtab, taxa_are_rows=FALSE))

input = merge_phyloseq(ps,map) 

input

microbiome::summarize_phyloseq(input)

```

## Subsetting and organizing data

```{r}

#Rarefaction
inputR = rarefy_even_depth(input, sample.size = 82270, replace = FALSE)

#Subsetting
Cedrela  = subset_samples(inputR, Tree == "Cedrela_fissilis")
Pelto    = subset_samples(inputR, Tree == "Peltophorum_dubium")
Cecropia = subset_samples(inputR, Tree == "Cecropia_pachystachya")
Bulk     = subset_samples(inputR, Tree == "Bulk")

#Creating clr objects
inputR.clr = microbiome::transform(inputR, "clr")

```
##Creating the R6 objects

```{r}

## for Cedrela ##

otu_ced = as.data.frame(t(otu_table(Cedrela)))
tax_ced = as.data.frame(tax_table(Cedrela))
map_ced = data.frame(sample_data(Cedrela))

dataset.ced <- microtable$new(sample_table = map_ced,
                              otu_table    = otu_ced,
                              tax_table    = tax_ced)
dataset.ced$tidy_dataset()
dataset.ced

## for Peltophorum ##

otu_pel = as.data.frame(t(otu_table(Pelto)))
tax_pel = as.data.frame(tax_table(Pelto))
map_pel = data.frame(sample_data(Pelto))

dataset.pel <- microtable$new(sample_table = map_pel,
                              otu_table    = otu_pel,
                              tax_table    = tax_pel)
dataset.pel$tidy_dataset()
dataset.pel

## for Cecropia ##

otu_cec = as.data.frame(t(otu_table(Cecropia)))
tax_cec = as.data.frame(tax_table(Cecropia))
map_cec = data.frame(sample_data(Cecropia))

dataset.cec <- microtable$new(sample_table = map_cec,
                              otu_table    = otu_cec,
                              tax_table    = tax_cec)
dataset.cec$tidy_dataset()
dataset.cec

## for bulk ##

otu_bk = as.data.frame(t(otu_table(Bulk)))
tax_bk = as.data.frame(tax_table(Bulk))
map_bk = data.frame(sample_data(Bulk))

dataset.bk <- microtable$new(sample_table  = map_bk,
                              otu_table    = otu_bk,
                              tax_table    = tax_bk)
dataset.bk$tidy_dataset()
dataset.bk

## for the whole dataset ##

otu_tp = as.data.frame(t(otu_table(inputR)))
tax_tp = as.data.frame(tax_table(inputR))
map_tp = data.frame(sample_data(inputR))

# Let's create a microtable object with more information
dataset <- microtable$new(sample_table = map_tp,
                          otu_table    = otu_tp,
                          tax_table    = tax_tp)
dataset

dataset$tax_table %<>% tidy_taxonomy

```

## FIGURE: Alpha Diversity

```{r}

#Calculating Alpha Diversity
observed=microbiome::alpha(inputR, index = "all")
meta=microbiome::meta(inputR)

#Creating a file to plot a graph
alpha= cbind(observed,meta)

alpha$Treatment <- factor(alpha$Treatment,
                          levels = c("Initial_AS", "Initial_ADE",
                                     "Cecropia_AS", "Cecropia_AS+ADE", "Cecropia_ADE",
                                     "Peltophorum_AS", "Peltophorum_AS+ADE",
                                     "Peltophorum_ADE",
                                     "Cedrela_AS, Cedrela_AS+ADE", "Cedrela_ADE"))

#Testing differences
dunn.test::dunn.test(alpha$observed, alpha$Treatment,
                     method="bonferroni", kw=TRUE, label=TRUE,
                     wrap=FALSE, table=F, list=T, rmc=FALSE, 
                     alpha=0.05, altp=FALSE)

comparison = list(c("Initial_AS", "Cecropia_AS"))

alpha.plot = 
  ggplot(data = alpha, aes(x=Treatment, y=observed, fill = Substrate)) +
  geom_boxplot() +
  labs(x = "Treatment", y= "Observed Values") +
  geom_jitter(shape=16, position=position_jitter(0)) +
  theme(axis.text.x = element_text(angle = -30, vjust = 0.5, hjust=0.5)) +
  stat_compare_means(comparisons = comparison) +
  labs(color='Treatment') + guides(color = "none")+
  scale_fill_manual(values = c("#7aa457", "#9e6ebd", "#cb6751"))
  
alpha.plot

#Don't forget setwd
#dev.print(tiff, "alpha_div.tiff", compression = "lzw", res=600, height=5, width=9, units="in")

```


## FIGURE: Beta Diversity

```{r}

#Firstly, a perMANOVA

df        = as(sample_data(inputR.clr), "data.frame")
ds        = phyloseq::distance(inputR.clr, method = "euclidean")
permanova = adonis2(ds ~ Treatment, data = df, permutations = 999)

permanova

#And the plot

input_ord = ordinate(inputR.clr, "PCoA" , "euclidean") 
p3 = plot_ordination(inputR.clr, input_ord, color = "Tree", shape = "Substrate")
p1 = p3 + geom_point(aes(shape = Substrate), size = 6, alpha = 0.8) +
  theme(legend.position = "right") +
  labs(title = "R-Squared = 0,30; p = 0.01") +
  scale_fill_manual(values = ze)

p1

#dev.print(tiff, "Beta_div.tiff", compression = "lzw", res=600, height=5, width=8, units="in")

```

## FIGURE: Distribution of phyla

```{r}

# create trans_abund object
# using 10 Phyla with the highest abundance in the dataset.
dataset$cal_abund()
t1 <- trans_abund$new(dataset = dataset, taxrank = "Phylum", ntaxa = 10)

# two facets example
# require package ggh4x, first run install.packages("ggh4x") if not installed
t1$plot_bar(others_color = "grey70", facet = "Tree",
            facet2 = "Substrate", xtext_keep = FALSE, legend_text_italic = FALSE, barwidth = 1)

#dev.print(tiff, "Beta_div.tiff", compression = "lzw", res=600, height=5, width=8, units="in")

```

## FIGURE: Production aspects

```{r}
#Excluding initial samples
#DM is the object with only T1 samples

DM <- filter(map_tpsa, U_brizantha_weight != "NA")
DM = DM %>%
  arrange(U_brizantha_weight) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(Substrate=factor(Substrate, levels=c("AS", "AS+ADE", "ADE")))

comparison = list(c("AS", "ADE"), c("AS", "AS+ADE"), c("AS+ADE", "ADE"))

## Urochloa brizantha ##

DM.plot = ggplot(data = DM, mapping = aes(x = Substrate, y = U_brizantha_weight, fill = Substrate)) +
  geom_boxplot() +
  theme(legend.position = "none") +
  stat_compare_means(comparisons = comparison) +
  labs(title = "U. brizantha production", y = "Dry mass (g)") +
  scale_fill_manual(values = c("#7aa457", "#9e6ebd", "#cb6751"))
DM.plot

#dev.print(tiff, "Urochloa.tiff", compression = "lzw", res=600, height=4, width=6, units="in")

```


```{r}

## Plants' dry matter ##
TM.plot = ggplot(data = DM, mapping = aes(x = Substrate, y = Dry_mass_production, fill = Treatment)) +
  geom_boxplot() +
  facet_grid(.~Tree, scales="free", space="free_x") +
  theme(legend.position="none") +
  labs(title="Dry matter production",
        x ="", y = "DM production (g)")

## Plant height ##
H.plot = ggplot(data = DM, mapping = aes(x = Substrate, y = Height, fill = Treatment)) +
  geom_boxplot() +
  facet_grid(.~Tree, scales="free", space="free_x") + 
  theme(legend.position="none") +
  labs(title="Plant height",
        x ="", y = "Height (cm)")

## Root size ##
RS.plot = ggplot(data = DM, mapping = aes(x = Substrate, y = Root_size, fill = Treatment)) +
  geom_boxplot() +
  facet_grid(.~Tree, scales="free", space="free_x") + 
  theme(legend.position="none") +
  labs(title="Root size",
        x ="Substrate", y = "Root size (cm)")

#Binding graphics
plants.g = ggarrange(TM.plot, H.plot, RS.plot, ncol=1, nrow = 3, align = "hv", common.legend = F)
plants.g

#dev.print(tiff, "Growing.tiff", compression = "lzw", res=600, height=8, width=6, units="in")

```
## FIGURE: Soil Chemistry

```{r}

t1 <- trans_env$new(dataset = dataset, add_data = map_tp[, c(11:17, 24:27, 35:38)])

t1$cal_diff(method = "KW_dunn", group = "Substrate")
# place all the plots into a list
tmp <- list()
for(i in colnames(t1$data_env)){
  tmp[[i]] <- t1$plot_diff(measure = i, add_sig_text_size = 5, xtext_size = 12) + theme(plot.margin = unit(c(0.1, 0, 0, 1), "cm"))
}
gridExtra::arrangeGrob(grobs = tmp, ncol = 4)

t1$cal_autocor(group = "Substrate")

#dev.print(tiff, "Chemistry.tiff", compression = "lzw", res=600, height=20, width=20, units="in")

```

## FIGURE: RDA

```{r}

#Analysing
t1$cal_ordination(method = "RDA", taxa_level = "Family")
t1$trans_ordination(show_taxa = 10, adjust_arrow_length = TRUE,
                    max_perc_env = 2.5, max_perc_tax = 2.5,
                    min_perc_env = 0.9, min_perc_tax = 0.9)

#Plotting
t1$plot_ordination(plot_color = "Substrate",
                   plot_shape = "Tree", plot_type = c("point", "ellipse"))

#dev.print(tiff, "RDA.tiff", compression = "lzw", res=600, height=6, width=8, units="in")

```

## FIGURE: Heatmap

```{r}

## Cecropia map ##

dataset.cec$tax_table %<>% tidy_taxonomy
dataset.cec$cal_abund()
t1 <- trans_abund$new(dataset = dataset.cec, taxrank = "Family", ntaxa = 15)
cec.p = t1$plot_heatmap(facet = "Substrate", xtext_keep = FALSE, withmargin = FALSE)

## Peltophorum map ##

dataset.pel$tax_table %<>% tidy_taxonomy
dataset.pel$cal_abund()
t2 <- trans_abund$new(dataset = dataset.pel, taxrank = "Family", ntaxa = 15)
pel.p = t2$plot_heatmap(facet = "Substrate", xtext_keep = FALSE, withmargin = FALSE)

## Cedrela map ##

dataset.ced$tax_table %<>% tidy_taxonomy
dataset.ced$cal_abund()
t3 <- trans_abund$new(dataset = dataset.pel, taxrank = "Family", ntaxa = 15)
ced.p = t3$plot_heatmap(facet = "Substrate", xtext_keep = FALSE, withmargin = FALSE)

## Bulk map ##

dataset.bk$tax_table %<>% tidy_taxonomy
dataset.bk$cal_abund()
t4 <- trans_abund$new(dataset = dataset.bk, taxrank = "Family", ntaxa = 15)
bk.p = t4$plot_heatmap(facet = "Substrate", xtext_keep = FALSE, withmargin = FALSE)

teste = ggarrange(bk.p, cec.p, pel.p, ced.p, ncol=2, nrow = 2, align = "hv", common.legend = T)
teste

#dev.print(tiff, "heatmap.tiff", compression = "lzw", res=600, height=6, width=8, units="in")

```

