---
title: "ADEs and Agricolae Soils"
author: "Anderson Freitas - Luis F. Zagatto"
date: "17/01/2022"
output: html_document
---

This is the R Markdown document for analysis of the project “Feedback planta-solo em áreas degradas por plantio de pastagens”.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Dirty work
Here we will put all the libraries we needed for downstream analysis. I will use it to simplify and unify all libraries in one only chunk.

```{r message=FALSE}
#Libraries

library(hues)
library(knitr)
library(readr)
library(vegan)
library(dplyr)
library(ggpubr)
library(ggplot2)
library(corrplot)
library(phyloseq)
library(agricolae)
library(tidyverse)
library(microbiome)
library(factoextra)

#############################################

summarySE <- function(data=NULL, measurevar, groupvars=NULL, na.rm=FALSE,
                      conf.interval=.95, .drop=TRUE) {
  library(plyr)
  
  # New version of length which can handle NA's: if na.rm==T, don't count them
  length2 <- function (x, na.rm=FALSE) {
    if (na.rm) sum(!is.na(x))
    else       length(x)
  }
  
  # This does the summary. For each group's data frame, return a vector with
  # N, mean, and sd
  datac <- ddply(data, groupvars, .drop=.drop,
                 .fun = function(xx, col) {
                   c(N    = length2(xx[[col]], na.rm=na.rm),
                     mean = mean   (xx[[col]], na.rm=na.rm),
                     sd   = sd     (xx[[col]], na.rm=na.rm)
                   )
                 },
                 measurevar
  )
  
  # Rename the "mean" column    
  datac <- plyr::rename(datac, c("mean" = measurevar))
  
  datac$se <- datac$sd / sqrt(datac$N)  # Calculate standard error of the mean
  
  # Confidence interval multiplier for standard error
  # Calculate t-statistic for confidence interval: 
  # e.g., if conf.interval is .95, use .975 (above/below), and use df=N-1
  ciMult <- qt(conf.interval/2 + .5, datac$N-1)
  datac$ci <- datac$se * ciMult
  
  return(datac)
}

#############################################

```


## Setting some questions of beauty

```{r}

theme_set(theme_bw())
pal = "Set3"
scale_colour_discrete <-  function(palname=pal, ...){
  scale_colour_brewer(palette=palname, ...)
}
scale_fill_discrete <-  function(palname=pal, ...){
  scale_fill_brewer(palette=palname, ...)
}

ze=c("#79b643",
"#bb59bd",
"#52ac7a",
"#cf426b",
"#4dadcf",
"#cf5234",
"#7475cb",
"#cea13d",
"#be6c91",
"#748037",
"#be774e")

```


## Starting with non-genomic data

At a first moment, we will add only the mapping file to proceed to initial ande descriptive analysis of the data.

```{r}

map_tpsa <- read_delim("E:/Anderson-BackUp/TP_SA/map_tpsa.tsv", delim = "\t",
                       escape_double = FALSE, na = "NA", trim_ws = TRUE)
map = data.frame(map_tpsa)

```


## Testing production of Brachiaria.
We grew Urochloa brizantha in all soil types. Plant number was normalized by the the vase with the minimal plant number. The results of dry matter production in different soils are shown here.

```{r}
#Excluding initial samples
#DM is the object with only T1 samples

DM <- filter(map, U_brizantha_weight != "NA")
DM = DM %>%
  arrange(U_brizantha_weight) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(Substrate=factor(Substrate, levels=c("AS", "AS+ADE", "ADE")))

```


## Is dry mass production of brachiaria different among substrates?

```{r}

dunn.test::dunn.test(DM$U_brizantha_weight, DM$Substrate, method="bonferroni", kw=TRUE, label=TRUE, 
      wrap=FALSE, table=F, list=T, rmc=FALSE, alpha=0.05, altp=FALSE)

```

Everythig is different. It should be placed in a graph.

```{r}
#DM graph

comparison = list(c("AS", "ADE"), c("AS", "AS+ADE"), c("AS+ADE", "ADE"))
DM.plot = ggplot(data = DM, mapping = aes(x = Substrate, y = U_brizantha_weight, fill = Substrate)) +
  geom_boxplot() +
  stat_compare_means(comparisons = comparison) +
  theme(legend.position = "none") 
DM.plot

```

What can we see? There is much more brachiaria in ADE and the introduction of some ADEs also increases the ability growing pasture in the agricolae soil.

## Is the dry mass production of trees different among substrates?

After the brachiaria withdrawal, we planted some species of plants in all soils. The results for mass production are here:

```{r}

TM = DM %>%
  arrange(Dry_mass_production) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(Treatment=factor(Treatment, levels=c("Cecropia_AS", "Cecropia_AS+ADE", "Cecropia_ADE",
                                              "Peltophorum_AS", "Peltophorum_AS+ADE", "Peltophorum_ADE",
                                              "Cedrela_AS", "Cedrela_AS+ADE", "Cedrela_ADE"
                                              )))
TM[is.na(TM)] <- 0

dunn.test::dunn.test(TM$Dry_mass_production, TM$Treatment, method="bonferroni", kw=TRUE, label=TRUE, 
      wrap=FALSE, table=F, list=T, rmc=FALSE, alpha=0.05, altp=FALSE)

```

Peltophorum presented the most promising data. Let's see it graphically.

```{r}
#Tree Matter graph

compar = list(c("Peltophorum_ADE", "Cedrela_ADE"), c("Cecropia_ADE", "Peltophorum_ADE"), c("Cecropia_ADE", "Cedrela_ADE"))

TM.plot = ggplot(data = TM, mapping = aes(x = Substrate, y = Dry_mass_production, fill = Treatment)) +
  geom_boxplot() +
  stat_compare_means(comparisons = compar) +
  facet_grid(.~Tree, scales="free", space="free_x") +
  theme(legend.position="none") +
  labs(title="Dry matter production",
        x ="", y = "DM production (g)")
TM.plot
  
```
The increased grow in all plants when cultivated with more ADE is very clear, even in adverse conditions. However, there is no significant difference among all plants in ADE only cultivation. 

## Another descriptive analyses for plants
Besides dry matter, we also measured height and root size of the plants. Results were as follows:

```{r}
#Height

dunn.test::dunn.test(TM$Height, TM$Treatment, method="bonferroni", kw=TRUE, label=TRUE, 
      wrap=FALSE, table=F, list=T, rmc=FALSE, alpha=0.05, altp=FALSE)
```
A lot of differences. But can we compare plant-to-plant?

```{r}
#Roots

dunn.test::dunn.test(TM$Root_size, TM$Treatment, method="bonferroni", kw=TRUE, label=TRUE, 
      wrap=FALSE, table=F, list=T, rmc=FALSE, alpha=0.05, altp=FALSE)

```

No differences in root size over substrates.

And graphically...

```{r}

#Graph for Height

compar = list(c("Peltophorum_ADE", "Cedrela_ADE"), c("Cecropia_ADE", "Peltophorum_ADE"), c("Cecropia_ADE", "Cedrela_ADE"))

H.plot = ggplot(data = TM, mapping = aes(x = Substrate, y = Height, fill = Treatment)) +
  geom_boxplot() +
  stat_compare_means(comparisons = compar) +
  facet_grid(.~Tree, scales="free", space="free_x") + 
  theme(legend.position="none") +
  labs(title="Plant height",
        x ="", y = "Height (cm)")
H.plot

```
```{r}
#Graph for Root size

compar = list(c("Peltophorum_ADE", "Cedrela_ADE"), c("Cecropia_ADE", "Peltophorum_ADE"), c("Cecropia_ADE", "Cedrela_ADE"))

RS.plot = ggplot(data = TM, mapping = aes(x = Substrate, y = Root_size, fill = Treatment)) +
  geom_boxplot() +
  stat_compare_means(comparisons = compar) +
  facet_grid(.~Tree, scales="free", space="free_x") + 
  theme(legend.position="none") +
  labs(title="Root size",
        x ="Substrate", y = "Root size (cm)")
RS.plot
```

As much as we have more ADE we have more height plants, but no bigger roots. It could be explained by strees being only thermical and no hydrical.

## Putting trees graphs together
We will unify those three graphs (Dry matter, height and root size) in one figure. Check this out:

```{r}
#Binding graphics

plants.g = ggarrange(TM.plot, H.plot, RS.plot, ncol=1, nrow = 3, align = "hv", common.legend = F)
plants.g

```
Now we have an interesting figure.

## Physical and Chemical properties of the soil
Here, we will analyse how the characteristics from soils influence and are influenced by chemical and physical analysis. We will show the data comparing the same soils over the time.

```{r}
#Setting data in the right order
QPh = mutate(map_tpsa, Treatment=factor(Treatment, levels=c("Initial_AS", "Cecropia_AS", "Peltophorum_AS", "Cedrela_AS",
                                              "Initial_ADE", "Cecropia_ADE", "Peltophorum_ADE", "Cedrela_ADE",
                                              "Cecropia_AS+ADE",  "Peltophorum_AS+ADE", "Cedrela_AS+ADE"
                                              )))
QPh

```

## What about pH among substrates and plants?

```{r}

dunn.test::dunn.test(map_tpsa$pH, map_tpsa$Substrate, method="bonferroni", kw=TRUE, label=TRUE, 
      wrap=FALSE, table=F, list=T, rmc=FALSE, alpha=0.05, altp=FALSE)

dunn.test::dunn.test(map_tpsa$pH, map_tpsa$Treatment, method="bonferroni", kw=TRUE, label=TRUE, 
      wrap=FALSE, table=F, list=T, rmc=FALSE, alpha=0.05, altp=FALSE)

```

It seems pH vary a little bit. Let's see the graph.

```{r}
#Plotting pH graph

pH.plot = ggplot(data = QPh, mapping = aes(x = Treatment, y = pH, fill = Treatment, color = Substrate)) +
  geom_point(size = 4) +
  labs(title="pH",
        x ="Treatment", y = "pH") +
  theme(legend.position="none") +
  scale_fill_manual(values=ze) +
  theme(axis.text.x = element_text(angle = 45))
pH.plot

```

But maybe it will be better show it in a table...

Firstly we will catch means and sd for all groups in all variables

```{r}
#Calculating the mean

mean.tab <- data.frame(QPh %>% 
  group_by(Treatment) %>%
  summarise(across(-Tree, mean, na.rm = TRUE)))

mean.tab

var.m = colnames(QPh[11:38])
var.m

num.vars <- QPh[,11:38]
num.vars

Treatment <- QPh$Treatment

kusk <- cbind(num.vars, Treatment)
kusk

colnames(kusk)

for (i in 1:28) {
  abc <- agricolae::kruskal(y = kusk[,i], trt = kusk$Treatment,
                      alpha=0.05, p.adj = "fdr", group = TRUE)
  print(abc)
}

```
We got a result. But I need to go back here to export this for a text file.
Or maybe exclude some colummns.

## Correlation among chemical/physical variables and growing
No extra explanation needed. LOL.

```{r}
#Separating only numerical variables

U.brizantha = DM$U_brizantha_weight
Tree.info   = DM[,7:9]
colnames(Tree.info) = c("Plant mass", "Plant height", "Root size")
rest        = DM[,-c(1:10)]
rest2        = rest[, -c(7,18:24)]
  
cor.var     = cbind(U.brizantha, Tree.info, rest2)

cor.var[is.na(cor.var)] <- 0                        #Transforming NAs in zeros

M           = cor(cor.var)                          #Calculating correlations
testRes     = cor.mtest(cor.var, conf.level = 0.95) #Obtaining p-values

p.val       = testRes$p
p.values     = p.val[c(1:4), -c(1:4)]

cor.good    = M[c(1:4), -c(1:4)]

corrplot(cor.good, p.mat = p.values, method = 'circle',
         insig='blank', number.cex = 0.8, tl.srt = 45)  #plotting correlations             

```
We found very interesting stuff, such as higher correlations of U. brizantha grouwing related to higher levels of carbon, organic matter, P, Ca, H, H:Al3, SB, CTC, Cu, Fe, Ca:CEC, and sand.

## Can we separate chemichals and plant characteristics by PCA?

```{r}
#Doing a PCA biplot
active <- map_tpsa[9:44, c(11:17,19:27,35:38)]  
active[is.na(active)] <- 0
colnames(active)[10]  <- "V%"

res.pca <- prcomp(active, scale = TRUE)

##per tretament

groups <- as.factor(map_tpsa$Treatment[9:44])
fviz_pca_biplot(res.pca,
             col.ind = groups, # color by groups
             geom = "point",
             palette = ze,
             addEllipses = TRUE, # Concentration ellipses
             ellipse.type = "confidence",
             legend.title = "Groups",
             repel = TRUE
)

##per substrate

groups <- as.factor(map_tpsa$Substrate[9:44])
fviz_pca_biplot(res.pca,
             col.ind = groups, # color by groups
             geom = "point",
             palette = ze,
             addEllipses = TRUE, # Concentration ellipses
             ellipse.type = "confidence",
             legend.title = "Groups",
             repel = TRUE
)

##per tree

groups <- as.factor(map_tpsa$Tree[9:44])
fviz_pca_biplot(res.pca,
             col.ind = groups, # color by groups
             geom = "point",
             palette = ze,
             addEllipses = TRUE, # Concentration ellipses
             ellipse.type = "confidence",
             legend.title = "Groups",
             repel = TRUE
)



#dev.print(tiff, "corr.tiff", compression = "lzw", res=600, height=6, width=8, units="in")

```

It seems that the substrates determinate everything.

## Importing files from dada2 and creating a phyloseq object
# It's NGS, baby

```{r}

seqtab = readRDS("dada2/seqtab_final.rds")
taxa = readRDS("dada2/taxonomy_genera.rds")
mapfile <- "E:/Anderson-BackUp/TP_SA/map_tpsa.csv"

map = import_qiime_sample_data(mapfile)

# Building a phyloseq object

ps <- phyloseq(tax_table(taxa), otu_table(seqtab, taxa_are_rows=FALSE))

input = merge_phyloseq(ps,map) 

input

microbiome::summarize_phyloseq(input)

```
## Alpha diversity measures

```{r}

#Alpha Diversity

inputR = rarefy_even_depth(input, sample.size = 82270, replace = FALSE)


#Calculating Alpha Diversity

observed=microbiome::alpha(inputR, index = "all")
meta=microbiome::meta(inputR)

#Creating a file to plot a graph
alpha= cbind(observed,meta)

alpha3 = filter(alpha, Period == "T1")

alpha2 = alpha %>%
  arrange(observed) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(Treatment=factor(Treatment, levels=c("Initial_AS", "Initial_ADE", "Cecropia_AS",
                                              "Cecropia_AS+ADE", "Cecropia_ADE",  "Peltophorum_AS",
                                              "Peltophorum_AS+ADE", "Peltophorum_ADE",
                                              "Cedrela_AS", "Cedrela_AS+ADE", "Cedrela_ADE"
                                              )))

alpha.plot = 
  ggplot(data = alpha2, aes(x=Treatment, y=observed, fill = Substrate)) +
  geom_boxplot() +
  labs(x = "Treatment", y= "Observed Values") +
  geom_jitter(shape=16, position=position_jitter(0)) +
  stat_compare_means(comparisons = comparison) +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=0.5)) +
  labs(color='Treatment') + guides(color = "none")+
  scale_fill_manual(values = c("#7aa457", "#9e6ebd", "#cb6751"))
  
alpha.plot


```

## Beta Diversity plot

First in NMDS

```{r}

#NMDS

input.clr = microbiome::transform(input, "clr")

input_ord = ordinate(input.clr, "NMDS" , "euclidean") 
p3 = plot_ordination(input.clr, input_ord, color = "Tree")
p1 = p3 + geom_point(aes(shape = Substrate), size = 6, alpha = 0.8) +
  theme(legend.position = "right", legend.title = element_blank()) +
  scale_fill_manual(values = ze)

p1

```

And then PCoA

```{r}

#PCoA

input_ord2 = ordinate(input.clr, "PCoA" , "euclidean") 
p4 = plot_ordination(input.clr, input_ord2, color = "Tree")
p2 = p4 + geom_point(aes(shape = Substrate), size = 6, alpha = 0.8) +
  theme(legend.position = "right", legend.title = element_blank()) +
  scale_fill_manual(values = ze)

p2

```

PCoA has showed the dissimilarity clearer. 

## Validating beta diversity by perMANOVA

```{r}

df        = as(sample_data(input.clr), "data.frame")
ds        = phyloseq::distance(input.clr, method = "euclidean")
permanova = adonis2(ds ~ Treatment, data = df, permutations = 999)
permanova2 = adonis2(ds ~ Tree, data = df, permutations = 999)
permanova3 = adonis2(ds ~ Substrate, data = df, permutations = 999)

permanova
permanova2
permanova3

```



## Overall community

```{r}

# Orders by Treatment

orders     = microbiome::aggregate_rare(input, level = "Order", detection = 1/100, prevalence = 1/100)
top10      = names(sort(taxa_sums(orders), decreasing = T))[1:10]
overall    = transform_sample_counts(orders, function(OTU) OTU/sum(OTU))
overall2   = prune_taxa(top10, overall)
#plot.ov    = subset_taxa(ps.fam.GHC, Order != "Unknown")
plot_bar(overall2, x="Sample_ID", fill="Order") + facet_wrap(~Treatment, scales="free_x")

```

```{r}

# Orders by plant

orders.p     = microbiome::aggregate_rare(input, level = "Order", detection = 1/100, prevalence = 1/100)
top10.p      = names(sort(taxa_sums(orders.p), decreasing = T))[1:10]
overall.p    = transform_sample_counts(orders.p, function(OTU) OTU/sum(OTU))
overall2.p   = prune_taxa(top10.p, overall.p)
#plot.ov    = subset_taxa(ps.fam.GHC, Order != "Unknown")
plot_bar(overall2.p, x="Sample_ID", fill="Order") + facet_wrap(~Tree, scales="free_x")

```

```{r}

# Phyla by Tree

phyla.p     = microbiome::aggregate_rare(input, level = "Phylum", detection = 1/100, prevalence = 1/100)
top10.p      = names(sort(taxa_sums(phyla.p), decreasing = T))[1:10]
overall.p    = transform_sample_counts(phyla.p, function(OTU) OTU/sum(OTU))
overall2.p   = prune_taxa(top10.p, overall.p)
#plot.ov    = subset_taxa(ps.fam.GHC, Order != "Unknown")
plot_bar(overall2.p, x="Sample_ID", fill="Phylum") + facet_wrap(~Tree, scales="free_x")

```

```{r}

# Phyla by Substrate

phyla.p     = microbiome::aggregate_rare(input, level = "Phylum", detection = 1/100, prevalence = 1/100)
top10.p     = names(sort(taxa_sums(phyla.p), decreasing = T))[1:10]
overall.p   = transform_sample_counts(phyla.p, function(OTU) OTU/sum(OTU))
overall2.p  = prune_taxa(top10.p, overall.p)
#plot.ov    = subset_taxa(ps.fam.GHC, Order != "Unknown")
plot_bar(overall2.p, x="Sample_ID", fill="Phylum") + facet_wrap(~Substrate, scales="free_x")


```

```{r}

#Initial overview: ADE's in time zero and AS in time zero
# Overall composition of TP and SA soil

only.tpsa = subset_samples(input, Period == "T0")

phyla     = microbiome::aggregate_rare(only.tpsa, level = "Phylum", detection = 1/100, prevalence = 1/100)
top10     = names(sort(taxa_sums(phyla), decreasing = T))[1:10]
overall   = transform_sample_counts(phyla, function(OTU) OTU/sum(OTU))
top10.plot  = prune_taxa(top10, overall)

plot_bar(top10.plot, x="Sample_ID", fill="Phylum") + facet_wrap(~Substrate, scales="free_x")


```





## Differential abundance over plants

We'll test the differential abundance in a set of groups

```{r}

# Initial 


```

